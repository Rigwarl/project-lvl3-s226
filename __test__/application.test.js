import fs from 'fs';
import path from 'path';
import { html } from 'js-beautify';
import timer from 'timer-promise';
import axios from 'axios';

import run from '../src/js/application';

const initHtml = fs.readFileSync(path.join(__dirname, '../src/template.html')).toString();
const hexletFeed = fs.readFileSync(path.join(__dirname, './feeds/hexlet.xml')).toString();
const inputEvent = new Event('input', { bubbles: true });
const getTree = () => html(document.body.innerHTML);

beforeAll(() => {
  document.documentElement.innerHTML = initHtml;
  run();
});

jest.mock('axios');

test('form', async () => {
  const form = document.querySelector('#rss-form');
  const urlInput = form.querySelector('[data-selector="url-input"]');
  const submitBtn = form.querySelector('[data-selector="submit"]');

  // ready
  expect(getTree()).toMatchSnapshot();

  // invalid
  urlInput.value = 'invalid url';
  urlInput.dispatchEvent(inputEvent);
  expect(getTree()).toMatchSnapshot();

  // empty
  urlInput.value = '';
  urlInput.dispatchEvent(inputEvent);
  expect(getTree()).toMatchSnapshot();

  // valid
  urlInput.value = 'valid.url';
  urlInput.dispatchEvent(inputEvent);
  expect(getTree()).toMatchSnapshot();

  // disabled
  axios.get.mockRejectedValue();
  urlInput.value = 'valid.url';
  urlInput.dispatchEvent(inputEvent);
  submitBtn.click();
  expect(getTree()).toMatchSnapshot();

  // error
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  // loaded feed
  axios.get.mockResolvedValue({ data: hexletFeed });
  urlInput.value = 'valid.url';
  urlInput.dispatchEvent(inputEvent);
  submitBtn.click();
  await timer.start(100);
  expect(getTree()).toMatchSnapshot();

  // exists
  urlInput.value = 'valid.url';
  urlInput.dispatchEvent(inputEvent);
  submitBtn.click();
  expect(getTree()).toMatchSnapshot();
});

