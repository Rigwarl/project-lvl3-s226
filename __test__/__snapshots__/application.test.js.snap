// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form 1`] = `
"<div class=\\"container pt-md-5 pb-md-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <hr class=\\"my-4\\">
        <form class=\\"row\\" id=\\"rss-form\\">
            <div class=\\"form-group col-md-6\\">
                <label for=\\"rss-url\\">RSS URL</label>
                <input id=\\"rss-url\\" data-selector=\\"url-input\\" type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"site.com/rss\\">
                <div class=\\"invalid-feedback\\" data-selector=\\"url-error\\"></div>
            </div>
            <div class=\\"col-12\\">
                <button type=\\"submit\\" class=\\"btn btn-primary\\" data-selector=\\"submit\\">Load RSS</button>
            </div>
        </form>
    </div>

    <div class=\\"row\\" id=\\"rss-list\\"></div>
</div>

<div class=\\"modal fade\\" id=\\"rss-modal\\" tabindex=\\"-1\\">
    <div class=\\"modal-dialog modal-dialog-centered\\">
        <div class=\\"modal-content\\">
            <div class=\\"modal-header\\">
                <h5 class=\\"modal-title\\" data-selector=\\"title\\">Modal title</h5>
                <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\">
            <span>×</span>
          </button>
            </div>
            <div class=\\"modal-body\\" data-selector=\\"content\\">Modal content</div>
        </div>
    </div>
</div>"
`;

exports[`form 2`] = `
"<div class=\\"container pt-md-5 pb-md-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <hr class=\\"my-4\\">
        <form class=\\"row\\" id=\\"rss-form\\">
            <div class=\\"form-group col-md-6\\">
                <label for=\\"rss-url\\">RSS URL</label>
                <input id=\\"rss-url\\" data-selector=\\"url-input\\" type=\\"text\\" name=\\"url\\" class=\\"form-control is-invalid\\" placeholder=\\"site.com/rss\\">
                <div class=\\"invalid-feedback\\" data-selector=\\"url-error\\">Invalid URL</div>
            </div>
            <div class=\\"col-12\\">
                <button type=\\"submit\\" class=\\"btn btn-primary\\" data-selector=\\"submit\\">Load RSS</button>
            </div>
        </form>
    </div>

    <div class=\\"row\\" id=\\"rss-list\\"></div>
</div>

<div class=\\"modal fade\\" id=\\"rss-modal\\" tabindex=\\"-1\\">
    <div class=\\"modal-dialog modal-dialog-centered\\">
        <div class=\\"modal-content\\">
            <div class=\\"modal-header\\">
                <h5 class=\\"modal-title\\" data-selector=\\"title\\">Modal title</h5>
                <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\">
            <span>×</span>
          </button>
            </div>
            <div class=\\"modal-body\\" data-selector=\\"content\\">Modal content</div>
        </div>
    </div>
</div>"
`;

exports[`form 3`] = `
"<div class=\\"container pt-md-5 pb-md-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <hr class=\\"my-4\\">
        <form class=\\"row\\" id=\\"rss-form\\">
            <div class=\\"form-group col-md-6\\">
                <label for=\\"rss-url\\">RSS URL</label>
                <input id=\\"rss-url\\" data-selector=\\"url-input\\" type=\\"text\\" name=\\"url\\" class=\\"form-control is-valid\\" placeholder=\\"site.com/rss\\">
                <div class=\\"invalid-feedback\\" data-selector=\\"url-error\\"></div>
            </div>
            <div class=\\"col-12\\">
                <button type=\\"submit\\" class=\\"btn btn-primary\\" data-selector=\\"submit\\">Load RSS</button>
            </div>
        </form>
    </div>

    <div class=\\"row\\" id=\\"rss-list\\"></div>
</div>

<div class=\\"modal fade\\" id=\\"rss-modal\\" tabindex=\\"-1\\">
    <div class=\\"modal-dialog modal-dialog-centered\\">
        <div class=\\"modal-content\\">
            <div class=\\"modal-header\\">
                <h5 class=\\"modal-title\\" data-selector=\\"title\\">Modal title</h5>
                <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\">
            <span>×</span>
          </button>
            </div>
            <div class=\\"modal-body\\" data-selector=\\"content\\">Modal content</div>
        </div>
    </div>
</div>"
`;

exports[`form 4`] = `
"<div class=\\"container pt-md-5 pb-md-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <hr class=\\"my-4\\">
        <form class=\\"row\\" id=\\"rss-form\\">
            <div class=\\"form-group col-md-6\\">
                <label for=\\"rss-url\\">RSS URL</label>
                <input id=\\"rss-url\\" data-selector=\\"url-input\\" type=\\"text\\" name=\\"url\\" class=\\"form-control is-invalid\\" placeholder=\\"site.com/rss\\">
                <div class=\\"invalid-feedback\\" data-selector=\\"url-error\\">Empty URL</div>
            </div>
            <div class=\\"col-12\\">
                <button type=\\"submit\\" class=\\"btn btn-primary\\" data-selector=\\"submit\\">Load RSS</button>
            </div>
        </form>
    </div>

    <div class=\\"row\\" id=\\"rss-list\\"></div>
</div>

<div class=\\"modal fade\\" id=\\"rss-modal\\" tabindex=\\"-1\\">
    <div class=\\"modal-dialog modal-dialog-centered\\">
        <div class=\\"modal-content\\">
            <div class=\\"modal-header\\">
                <h5 class=\\"modal-title\\" data-selector=\\"title\\">Modal title</h5>
                <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\">
            <span>×</span>
          </button>
            </div>
            <div class=\\"modal-body\\" data-selector=\\"content\\">Modal content</div>
        </div>
    </div>
</div>"
`;

exports[`form 5`] = `
"<div class=\\"container pt-md-5 pb-md-5\\">
    <div class=\\"jumbotron\\">
        <h1>RSS Reader</h1>
        <hr class=\\"my-4\\">
        <form class=\\"row\\" id=\\"rss-form\\">
            <div class=\\"form-group col-md-6\\">
                <label for=\\"rss-url\\">RSS URL</label>
                <input id=\\"rss-url\\" data-selector=\\"url-input\\" type=\\"text\\" name=\\"url\\" class=\\"form-control\\" placeholder=\\"site.com/rss\\" disabled=\\"\\">
                <div class=\\"invalid-feedback\\" data-selector=\\"url-error\\"></div>
            </div>
            <div class=\\"col-12\\">
                <button type=\\"submit\\" class=\\"btn btn-primary\\" data-selector=\\"submit\\" disabled=\\"\\">Load RSS</button>
            </div>
        </form>
    </div>

    <div class=\\"row\\" id=\\"rss-list\\"></div>
</div>

<div class=\\"modal fade\\" id=\\"rss-modal\\" tabindex=\\"-1\\">
    <div class=\\"modal-dialog modal-dialog-centered\\">
        <div class=\\"modal-content\\">
            <div class=\\"modal-header\\">
                <h5 class=\\"modal-title\\" data-selector=\\"title\\">Modal title</h5>
                <button type=\\"button\\" class=\\"close\\" data-dismiss=\\"modal\\">
            <span>×</span>
          </button>
            </div>
            <div class=\\"modal-body\\" data-selector=\\"content\\">Modal content</div>
        </div>
    </div>
</div>"
`;
